packages/flutter/lib/src/widgets/app_model.dart

```diff
@@ -0,0 +1,207 @@
+  /// allowed or necessary. The `SharedObject` example above
+  /// demonstrates this.
+  ///
+  /// Unlike [AppModel.setValue], this method does _not_ cause dependent widgets
+  /// to be rebuilt.
+  ///
+  /// The type parameter `K` is the type of the value's keyword and `V`
+  /// is the type of the value. The value type is nullable, for example
+  /// `AppModel.initValue<String, String>(key,value)` means that the key parameter
+  /// must non-null but the value parameter can be null or a string.
+  static void initValue<K extends Object, V>(BuildContext context, K key, V? value) {
+    final _AppModelData? model = context.findAncestorWidgetOfExactType<_AppModelData>();
+    assert(_debugHasAppModel(model, context, 'initValue'));
+    model!.appModelState.initValue<K, V>(key, value);
+  }
+
+  static bool _debugHasAppModel(_AppModelData? model, BuildContext context, String methodName) {
+    assert(() {
+      if (model == null) {
+        throw FlutterError.fromParts(
+          <DiagnosticsNode>[
+            ErrorSummary('No AppModel widget found.'),
+            ErrorDescription('AppModel.$methodName requires an AppModel widget ancestor.\n'),
+            context.describeWidget('The specific widget that could not find an AppModel ancestor was'),
+            context.describeOwnershipChain('The ownership chain for the affected widget is'),
+            ErrorHint(
+              'Typically, the AppModel widget is introduced by the MaterialApp '
+              'or WidgetsApp widget at the top of your application widget tree. It '
+              'provides a key/value map of data that is shared with the entire '
+              'application.',
+            ),
+          ],
+        );
+      }
+      return true;
+    }());
+    return true;
+  }
+}
+
+class _AppModelState extends State<AppModel> {
+  late Map<Object, Object?> data = <Object, Object?>{};
+
+  @override
+  Widget build(BuildContext context) {
+    return _AppModelData(appModelState: this, child: widget.child);
+  }
+
+  V? getValue<K extends Object, V>(K key) => data[key] as V?;
+
+  void setValue<K extends Object, V>(K key, V? value) {
+    if (data[key] != value) {
+      setState(() {
+        data = Map<Object, Object?>.from(data);
+        data[key] = value;
+      });
+    }
+  }
+
+  void initValue<K extends Object, V>(K key, V? value) {
+    data = Map<Object, Object?>.from(data);
+    data[key] = value;
+  }
+}
+
+class _AppModelData extends InheritedModel<Object> {
+  _AppModelData({
+    Key? key,
+    required this.appModelState,
+    required Widget child
+  }) : data = appModelState.data, super(key: key, child: child);
+
+  final _AppModelState appModelState;
+  final Map<Object, Object?> data;
+
+  @override
+  bool updateShouldNotify(_AppModelData old) {
+    return data != old.data;
+  }
+
+  @override
+  bool updateShouldNotifyDependent(_AppModelData old, Set<Object> keys) {
+    for (final Object key in keys) {
+      if (data[key] != old.data[key]) {
```

Comment:
What about keys that are removed/added?

Reply:
Actually since we can't remove keys atm this probably wouldn't be an issue I think